### PR TITLE
Create a frontend UI for users to specify quantum compilation pipelines 

### PR DESCRIPTION
**Pull request by paul0403**
**Context:** 
We wish to create a frontend `pipeline` function for users to specify what circuit transformation passes to run in the dictionary form.
https://app.shortcut.com/xanaduai/epic/67492/p1-a-ui-is-available-for-users-to-specify-quantum-compilation-pipelines-from-python

**Description of the Change:**
A new API decorator function `catalyst.passes.pipeline` can be applied to qnodes. The decorator takes in a dictionary specifying what circuit transformation passes to run on the qnode with what options, and adds the passes to the `transform_named_sequence`.

In addition, `qjit` now takes in a kwarg `circuit_transform_pipeline` which expects the same pipeline dictionary. This will apply the pipeline to all qnodes in the qjit.

To test the pipeline, we added the boilerplate for a merge_rotation pass with a pass option. The pass is currently empty and does nothing. The merge_rotation boilerplate pass is not exposed as a user-facing API, and only exists for the purpose of testing the pipeline.

**Benefits:**
User can specify pass pipelines.

**Possible Drawbacks:**
There are two items of improvements possible:
1. The target qnode of the pass is recorded by name. This is not optimal. The quantum scope work will likely put each qnode into a module instead of a `func.func ... attributes {qnode}`. When that is in place, the qnode's module can have a proper attribute (as opposed to discardable) that records its transform schedule, i.e.
```
module_with_transform @name_of_module {
  // transform schedule
} {
  // contents of the module
}
```
This eliminates the need for matching target functions by name.

2. The number of `qjit` kwargs is maybe too many. As of now we implement the syntax specified by the epic, but there could be an alternate design.

[sc-67520]

### CODE REVIEW

**Comment by paul0403** [comment]
@rmoyard The epic description says the zne mitigation pass should go into the `catalyst.passes` module. I wanted to keep the old usage so in `passes.py` I just created a wrapper that calls the underlying `mitigate_with_zne` (the currently existing one). 

The spirit of the epic is that the `passes.py` module should control what passes to run. Therefore I turned off `lower-mitigation` in the default pipeline and moved it into the `apply-transform-sequence` (see #911). The behavior is the same as before:
- for functions not under `mitigate_with_zne`, nothing happens. Originally because the pass in the default pipeline ignores them, now because the pass is not run at all
- for functions under `mitigate_with_zne`, the pass runs. Originally because the pass runs in the default pipeline, now because `mitigate_with_zne` registers the pass in the transform sequence.

Thoughts on this?

File: frontend/test/lit/test_peephole_optimizations.py [hunk 1]
```diff
@@ -75,6 +76,123 @@ def func():
...
 
+#
+# pipeline
+#
+
+
+def test_pipeline_lowering():
+    my_pipeline = {
+        "cancel_inverses": {},
+        "mitigate_with_zne": {
```
Comment (rmoyard) [inline frontend/test/lit/test_peephole_optimizations.py:87]: Not sure about having ZNE in the pipeline like that. ZNE is not only applying a MLIR pass, there is stuff happening in the frontend as well. Otherwise why not to put the gradients as well?  In the pipeline should it not be optimization passes only? Also I think we restrict the ZNE to be applied on QNode. 

To my it is not clear what is the scope of the pipeline? We already have transform program in PL which is very similar. I think the pipeline should be a list of MLIR passes to apply on the function and not a list of transforms.

Comment (paul0403) [reply frontend/test/lit/test_peephole_optimizations.py:87]: Having ZNE is indeed quite painful implementation-wise, but it was specified in the epic (see PR description).

@josh146 does zne really need to be included in the pipeline? Personally I agree with Romain, that the pipeline should be restricted to "ZX-calculus"-style transformations.

Comment (paul0403) [reply frontend/test/lit/test_peephole_optimizations.py:87]: If we don't have ZNE, then it would be nice to have the merge rotation pass so that the pipeline can be meaningfully tested with more than one pass (right now there's just cancel inverses).

An alternative can be a empty peephole pass that does nothing (or inserts an identity, just for testing's sake).

Comment (rmoyard) [reply frontend/test/lit/test_peephole_optimizations.py:87]: The merge rotation pass is planned for next quarter, so next iteration :+1:

**Comment by paul0403** [comment]
Marking ready for review to check CI.

Two remaining items:
1. Do we really need to include ZNE? @josh146 
2. Make local configurations override global configurations

File: frontend/test/pytest/test_peephole_optimizations.py [hunk 2]
```diff
@@ -91,6 +92,39 @@ def g(x):
     assert np.allclose(workflow()[0], workflow()[1])
 
 
+@pytest.mark.parametrize("theta", [42.42])
```
Comment (paul0403) [inline frontend/test/pytest/test_peephole_optimizations.py:95]: changelog

File: frontend/catalyst/qfunc.py [hunk 3]
```diff
@@ -109,9 +110,11 @@ def extract_backend_info(
 
     # pylint: disable=no-member
     @debug_logger
-    def __call__(self, *args, **kwargs):
+    def __call__(self, pass_pipeline, *args, **kwargs):
         assert isinstance(self, qml.QNode)
 
+        self = pipeline(pass_pipeline=pass_pipeline)(self)
```
Comment (paul0403) [inline frontend/catalyst/qfunc.py:116]: local pipeline need to override global pipeline

File: frontend/catalyst/jax_primitives.py [hunk 4]
```diff
@@ -528,16 +529,28 @@ def _apply_registered_pass_lowering(
         """
 
     if first_op_in_block.name == "transform.apply_registered_pass":
+        """
```
Comment (paul0403) [inline frontend/catalyst/jax_primitives.py:532]: This reverse ordering of jaxpr primitives and mlir aaply_registered_pass is only due to the fact that we are repeatedly wrapping.

If we exclude ZNE, then all passes will be of the same format (https://github.com/PennyLaneAI/catalyst/blob/main/mlir/include/Quantum/Transforms/Passes.td#L94), and the wrapping is unnecessary since we can just bind the apply_registered_pass primitive in the pipeline decorator, instead of having it call APIs.

File: frontend/catalyst/passes.py [hunk 5]
```diff
@@ -33,15 +33,69 @@
...
+from catalyst.tracing.contexts import EvaluationContext
 
 
 ## API ##
 # pylint: disable=line-too-long
-def cancel_inverses(fn=None):
+def pipeline(fn=None, *, pass_pipeline=None):
+    """
+    Here are documentation words
+    """
```
Comment (josh146) [inline frontend/catalyst/passes.py:52]: 😅

Comment (josh146) [reply frontend/catalyst/passes.py:52]: Feel free to copy the docstring directly from the epic, https://app.shortcut.com/xanaduai/epic/67492?group_by=none&vc_group_by=day&ct_workflow=all&cf_workflow=500000005, let me know if we need to make any changes from what I originally wrote

**Comment by mehrdad2m** [comment]
I have some thoughts/questions about the design of the pipeline. Something that is worth clearing up is the definition of a "pass". My understanding of a compiler optimization pass is a transformation from one representation of the code to another without changing its functionality with hope that the transformed version is more optimized (runs faster, better memory usage, ...). 
With this definition, I think there should be a distinction between things like cancel_inverses and mitigate_with_zne and only the former should be considered as a compiler optimization pass. 
My question is then would the user want to run a performance related pipeline (that includes things like cancel_inverses) on individual functions within one qjit call? 
My instinct is that the answer is mostly no in which case we can just tie the pipeline to the qjit (similar to how clang runs O2 or O3 on the whole module) instead which makes the implementation nicer and simpler. This would be something like the pipelines option of qjit except that instead of replacing the whole compilation pipeline, it modifies the list of QUANTUM_COMPILATION_PASS passes. 
In some specific cases that we want a function to go through a certain optimization pass, we can still use the individual decorators that would work similar to pragma hints in c++. Any thoughts?

If the answer to the question is yes then ignore this comment :)

**Comment by josh146** [comment]
> My question is then would the user want to run a performance related pipeline (that includes things like cancel_inverses) on individual functions within one qjit call?

Hey @mehrdad2m, this is a good question! My expectation is that yes, _certain_ users would want to apply different optimizations to different circuits within the same `@qjit` function -- namely, users who are investigating and researching quantum compilation.

File: frontend/catalyst/qfunc.py (hunk 3 above)
Comment (paul0403) [reply frontend/catalyst/qfunc.py]: done

File: frontend/catalyst/passes.py (hunk 5 above)
Comment (paul0403) [reply frontend/catalyst/passes.py]: done

File: frontend/catalyst/passes.py [hunk 6]
```diff
@@ -33,14 +33,141 @@
...
+                for option, option_value in pass_options.items():
+                    opt += " " + str(option) + "=" + str(option_value)
+                apply_registered_pass_p.bind(
+                    pass_name=pass_names[API_name],
+                    options=f"func-name={fn_original_name}" + "_transformed" + uniquer + opt,
+                )
+        return wrapped_qnode_function(*args, **kwrags)
+
+    fn_clone.func = wrapper
+    fn_clone._peephole_transformed = True  # pylint: disable=protected-access
```
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/passes.py:166]: This is for the global config to know whether the qnode already has a local config https://github.com/PennyLaneAI/catalyst/pull/1131/files#diff-6248f6a10c0364879c7d6f3c40b38db9d59a91a2ad9f06e8cf0fac34e0ccaaeaR120. 

I don't like this way either, but this is the least resistance method I could think of.

**Review by mehrdad2m — COMMENTED** [review]
Very nice work 🚀

File: frontend/catalyst/passes.py [hunk 7]
```diff
@@ -33,14 +33,141 @@
...
+        my_pass_pipeline = {
+            "cancel_inverses": {},
+            "merge_rotations": {},
+        }
+
+        @qjit(pipeline=my_pass_pipeline)
+        def fn(x):
+            return jnp.sin(circuit(x ** 2))
+
+    Global and local (via ``@pipeline``) configurations can coexist, however local pass pipelines
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/passes.py:129]: > Also on another note, would the pipeline option override the circuit_transform_pipeline option? Maybe it is worth adding a paragraph that explains their relation to avoid confusion.

Yes.

**Comment by paul0403** [comment]
Summary of results from discussion:

1. The currently scope of the transform dialect is "function passes that need no processing in the frontend". This excludes ZNE.

2. (future work) Devices can have their own default pipeline, i.e. as a device implementor, I want to run pass X, Y, Z on programs that target my device. Users don't need to set it themselves neither on the qjit, nor qnode

3. (future work) The quantum scope work will likely put each qnode into a module instead of a `func.func ... attributes {qnode}`. When that is in place, the qnode's module can have a proper attribute (as opposed to discardable) that records its transform schedule, i.e. 
```
module_with_transform @name_of_module {
  // transform schedule
} {
  // contents of the module
}
```
This eliminates the need for matching target functions by name.

File: frontend/catalyst/passes.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/passes.py:129]: > I think it can easily be confused with `circuit_transform_pipeline` since we are talking about global and local pipeline.

In the documentation the passes that can be used under `circuit_transform_pipeline` is explicitly specified: https://github.com/PennyLaneAI/catalyst/pull/1131/files#diff-648249b9fca379a495b52fc5c000fb7c4425daa46b6044df3ddf4515b3bc7b0aR143

[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/passes.py:129]: > Also on another note, would the `pipeline` option override the `circuit_transform_pipeline` option? Maybe it is worth adding a paragraph that explains their relation to avoid confusion.

All `pipeline` can control is whether `apply-transform-sequence` is run or not. The `circuit_transform_pipeline` controls what is in the transform sequence.

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -53,6 +53,67 @@
   Array([[1], [0], [1], [1], [0], [1],[0]], dtype=int64))
   ```
 
+* A new function `catalyst.passes.pipeline` allows the quantum circuit transformation pass pipeline for QNodes within a qjit-compiled workflow to be configured.
+  [(#1131)](https://github.com/PennyLaneAI/catalyst/pull/1131)
+
+  ```python
+    my_pass_pipeline = {
```
Comment (rmoyard) [inline doc/releases/changelog-dev.md:60]:
```suggestion
    my_passes = {
```

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -53,6 +53,67 @@
...
+        "cancel_inverses": {},
+        "merge_rotations": {},
+    }
+    dev = qml.device("lightning.qubit", wires=2)
+
+    @pipeline(my_pass_pipeline)
```
Comment (rmoyard) [inline doc/releases/changelog-dev.md:66]:
```suggestion
    @pipeline(my_passes)
```

File: frontend/catalyst/compiler.py [hunk 10]
```diff
@@ -77,6 +77,10 @@ class CompileOptions:
         experimental_capture (bool): If set to ``True``,
             use PennyLane's experimental program capture capabilities
             to capture the function for compilation.
+        circuit_transform_pipeline (Optional[dict[str, dict[str, str]]]):
```
Comment (rmoyard) [inline frontend/catalyst/compiler.py:80]: No action here, but @josh146 @isaacdevlugt this will soon beat the number of qnode's arguments.

File: frontend/catalyst/jit.py [hunk 11]
```diff
@@ -623,6 +637,7 @@ def closure(qnode, *args, **kwargs):
                 kwargs,
             )
 
+        catalyst.passes.PipelineNameUniquer.reset()
```
Comment (rmoyard) [inline frontend/catalyst/jit.py:640]: What does it do? Can you import it?

File: mlir/lib/Quantum/Transforms/merge_rotation.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+#define GEN_PASS_DEF_MERGEROTATIONPASS
+#define GEN_PASS_DECL_MERGEROTATIONPASS
+#include "Quantum/Transforms/Passes.h.inc"
+
+struct MergeRotationPass : impl::MergeRotationPassBase<MergeRotationPass> {
+    using MergeRotationPassBase::MergeRotationPassBase;
+
+    void runOnOperation() final
+    {
```
Comment (rmoyard) [inline mlir/lib/Quantum/Transforms/merge_rotation.cpp:38]: Can you make it such that it does nothing? It is possible that users will be tempted to try it

**Review by rmoyard — COMMENTED** [review]
Thanks, here are some comments

File: mlir/lib/Quantum/Transforms/merge_rotation.cpp (hunk 12 above)
Comment (paul0403) [reply mlir/lib/Quantum/Transforms/merge_rotation.cpp:38]: I made it so that we can test the pass options are passed down from frontend to mlir correctly. In added pytest there's a check for the message. Without that there would be no tests for this. I think this is fine as `merge_rotation` will not be added to the frontend documentation in this PR.

File: frontend/catalyst/jit.py (hunk 11 above)
Comment (paul0403) [reply frontend/catalyst/jit.py:640]: This is for the following case:

```python3
my_pipeline = {...}
my_other_pipeline = {...}

@qml.qnode(dev)
def f(x):
   ...

f_pipeline = pipeline(my_pipeline)(f)

@qjit(circuit_transform_pipeline=my_other_pipeline)
def workflow():
    return f(1.2), f_pipeline(1.2)
```

Without the name uniquer both the local and global pipeline configurations will produce a function with the name `f_transformed`.

Comment (paul0403) [reply frontend/catalyst/jit.py]: I have changed it to an import.

File: mlir/lib/Quantum/Transforms/merge_rotation.cpp (hunk 12 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/merge_rotation.cpp:38]: You can also create a test-pass instead of one that has a specific name that actually means something. :) That way, users wouldn't care about running a test-pass.

Comment (paul0403) [reply mlir/lib/Quantum/Transforms/merge_rotation.cpp:38]: > You can also create a test-pass instead of one that has a specific name that actually means something

I think we are against having "empty code" in the repo, e.g. second point of [this comment](https://github.com/PennyLaneAI/catalyst/pull/872#issuecomment-2237714359). 

I used merge rotation because it will soon become filled with actual content instead of being a true empty boilerplate. I don't want users to access it already so I did not add a `merge_rotation` frontend API to the documentation.

Comment (paul0403) [reply mlir/lib/Quantum/Transforms/merge_rotation.cpp:38]: I removed merge_rotation in all public documentation.
